Add option to start simulator via docker

Turnlings opened the pull request:
Adds the option to start the simulator via docker from a modal opened under the "Advanced" navbar item.

### Frontend
<img width="477" height="383" alt="image" src="https://github.com/user-attachments/assets/8155719b-8f8c-49a4-ab04-41a0a191a92e" />

Modal allows for configuration of port and vehicle type, aswell as checkbox for connecting after start. There is some easy potential for adding the other parameters from DEVELOPMENT_GUIDE.md, but was starting to suffer a bit from scope creep I thought.

### Features
- Set port(s)
- Set vehicle type (ArduCopter or ArduPlane)
- Shows loading notification if the docker container has to be downloaded for the first time
- Connect after start by waiting for "YOU CAN NOW CONNECT" message
- Lots of error messages for docker exceptions

### Notes
- This feature is primarily developmental so will NOT be tested
- This feature must be enabled under the "Experimental Developer Features" setting

Turnlings commented on gcs/src/redux/middleware/socketMiddleware.js:
```diff
@@ -420,6 +434,59 @@ const socketMiddleware = (store) => {
+            )
+            simulationLoadingId = null
+          }
+        })
```
Never gonna happen, not worth the time to implement

Turnlings commented on radio/app/endpoints/simulation.py:
```diff
@@ -0,0 +1,372 @@
+            if time.time() > deadline:
+                failure_reason = f"Container did not become ready within {timeout}s"
+                break
+
```
I have considered this and implemented in many different ways that you have suggested, each of which you have subsequently had a problem with. So I have settled on this, the simplest version, for maintainabilities sake.

Turnlings commented on gcs/src/redux/slices/simulationParamsSlice.js:
```diff
@@ -0,0 +1,99 @@
+      state.simulationStatus = action.payload
+    },
+    addSimulationPort: (state) => {
+      state.ports.push({ hostPort: 5760, containerPort: 5760 })
```
It's good enough as it is

Turnlings commented on radio/app/endpoints/simulation.py:
```diff
@@ -0,0 +1,375 @@
+            remove=True,
+            command=cmd,
+        )
+
```
This would be handled properly anyway

Turnlings commented on gcs/src/redux/middleware/socketMiddleware.js:
```diff
@@ -420,6 +434,73 @@ const socketMiddleware = (store) => {
+          msg.success
+            ? showSuccessNotification(msg.message)
+            : showErrorNotification(msg.message)
+
```
This pattern is used everywhere else in the codebase no issues

Turnlings commented on radio/app/endpoints/simulation.py:
```diff
@@ -0,0 +1,415 @@
+        emit_error_message(e.user_message)
+        if e.original_exception:
+            logger.exception(e)
+        return
```
No it wont because if it raises an error it will return

NexInfinite reviewed (dismissed):
Code looks good, copilot is going crazy

1Blademaster commented on radio/app/endpoints/simulation.py:
```diff
@@ -0,0 +1,419 @@
+    try:
+        validated_ports, primary_host_port = validate_ports(data.get("ports"))
+
+        connect = data["connect"] if "connect" in data else False
```
You can replace this with `connect = data.get('connect', False)`

1Blademaster reviewed (changes requested):
I'll test locally tomorrow, but adding these comments for now.

Turnlings commented on radio/app/endpoints/simulation.py:
```diff
@@ -0,0 +1,419 @@
+        ) from e
+
+
+def cleanup_container(container: Any) -> None:
```
I thought that might cause confusion with `stop_docker_simulation` as both technically stop the container, `stop_docker_simulation` being the for the route from the ui, and this one `cleanup_container` being used for just that, cleanup.

1Blademaster commented on gcs/src/components/toolbar/simulationModal.jsx:
```diff
@@ -0,0 +1,246 @@
+          />
+
+          <NumberInput
+            label="Container Port"
```
Nvm I had it other way around my bad 🤦 ignore me

1Blademaster commented:
Can we somehow show the loading/connection modal (which shows connection progress) while connecting to the simulation?

Turnlings commented:
> Can you swap this out for the trash icon, take a look at the mission item row for example

Done.

> Can we somehow show the loading/connection modal (which shows connection progress) while connecting to the simulation?

Extracted the progress bar at the bottom of the connect menu to its own component and added it to the bottom of the simulation modal.

> Got to add a disconnect from the simulation if stopping the simulation and are still connected

Done for if auto-connects after start, so will not work if simulation started and connected separately, but this will act the same as when container stopped from command line.

> Do we care about any ports other than 5760, 5762 and 5763? SITL doesn't open any others for access

The added ports now start from 5762 rather than 5761 so first 3 ports if no changes made will be 5760, 5762, 57623.

1Blademaster reviewed (approved):
Big PR. Diffs look good to me and initial tests went smoothly. I think in the future we can ship a custom dockerfile so we don't have to pull from the internet.
